services/error.go

```diff
@@ -73,88 +78,76 @@ type StorageError struct {
+func (e StorageError) Unwrap() error {
 	return e.Err
 }
 
-// NewMetadataNotRecognizedError will create a new MetadataUnrecognizedError.
-func NewMetadataNotRecognizedError(key string, value interface{}) *MetadataUnrecognizedError {
-	return &MetadataUnrecognizedError{
-		Err:   ErrCapabilityInsufficient,
-		Key:   key,
-		Value: value,
-	}
-}
-
 // MetadataUnrecognizedError means this operation meets unrecognized metadata.
 type MetadataUnrecognizedError struct {
-	Err error
-
 	Key   string
 	Value interface{}
 }
 
-func (e *MetadataUnrecognizedError) Error() string {
-	return fmt.Sprintf("metadata unrecognized, %s, %v: %s", e.Key, e.Value, e.Err.Error())
+func (e MetadataUnrecognizedError) Error() string {
+	return fmt.Sprintf("metadata unrecognized, %s, %v: %s", e.Key, e.Value, ErrCapabilityInsufficient.Error())
 }
 
 // Unwrap implements xerrors.Wrapper
-func (e *MetadataUnrecognizedError) Unwrap() error {
-	return e.Err
-}
-
-// NewPairUnsupportedError will create a new PairUnsupportedError.
-func NewPairUnsupportedError(pair types.Pair) *PairUnsupportedError {
-	return &PairUnsupportedError{
-		Err:  ErrCapabilityInsufficient,
-		Pair: pair,
-	}
+func (e MetadataUnrecognizedError) Unwrap() error {
+	return ErrCapabilityInsufficient
 }
 
 // PairUnsupportedError means this operation has unsupported pair.
 type PairUnsupportedError struct {
-	Err error
-
 	Pair types.Pair
 }
 
-func (e *PairUnsupportedError) Error() string {
-	return fmt.Sprintf("pair unsupported, %s: %s", e.Pair, e.Err.Error())
+func (e PairUnsupportedError) Error() string {
+	return fmt.Sprintf("pair unsupported, %s: %s", e.Pair, ErrCapabilityInsufficient.Error())
 }
 
 // Unwrap implements xerrors.Wrapper
-func (e *PairUnsupportedError) Unwrap() error {
-	return e.Err
-}
-
-// NewPairRequiredError will create a new PairRequiredError.
-func NewPairRequiredError(keys ...string) *PairRequiredError {
-	return &PairRequiredError{
-		Err:  ErrRestrictionDissatisfied,
-		Keys: keys,
-	}
+func (e PairUnsupportedError) Unwrap() error {
+	return ErrCapabilityInsufficient
 }
 
 // PairRequiredError means this operation has required pair but missing.
 type PairRequiredError struct {
-	Err error
-
 	Keys []string
 }
 
-func (e *PairRequiredError) Error() string {
-	return fmt.Sprintf("pair required, %v: %s", e.Keys, e.Err.Error())
+func (e PairRequiredError) Error() string {
+	return fmt.Sprintf("pair required, %v: %s", e.Keys, ErrRestrictionDissatisfied.Error())
 }
 
 // Unwrap implements xerrors.Wrapper
-func (e *PairRequiredError) Unwrap() error {
-	return e.Err
+func (e PairRequiredError) Unwrap() error {
+	return ErrRestrictionDissatisfied
+}
+
+// ObjectModeInvalidError means the provided object mode is invalid.
+type ObjectModeInvalidError struct {
+	Expected types.ObjectMode
+	Actual   types.ObjectMode
+}
+
+func (e ObjectModeInvalidError) Error() string {
+	return fmt.Sprintf("object mode invalid, expected %b, actual %b", e.Expected, e.Actual)
```

Comment:
`ObjectMode` && `ListMode` should implement `String() string` so that we can print human readable error here.